File: tests/Aether.Core.Tests/SignalFixtureTests.cs
// SPDX-License-Identifier: MIT
//
// Verifies that the C# Signal-Protocol crypto produces the same byte-level
// outputs that fixtures/signal/expected/*.json record. These same fixtures
// are consumed by every other language implementation, so any drift between
// languages surfaces as a failing test in whichever language diverges.
//
// To regenerate the fixtures (i.e. you intentionally changed the protocol),
// set the env var AETHER_SIGNAL_FIXTURES_REGEN=1 — the test prints expected
// values to stdout instead of asserting. Copy the printed values into the
// committed fixtures/signal/expected/*.json files, then re-verify every
// language. Any change is a wire-break event.

using System.Reflection;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using AetherMesh.Security.Services;
using Org.BouncyCastle.Crypto.Parameters;
using Xunit;
using Xunit.Abstractions;

namespace AetherMesh.Core.Tests;

public class SignalFixtureTests
{
    private readonly ITestOutputHelper _output;
    private static bool RegenerateMode =>
        Environment.GetEnvironmentVariable("AETHER_SIGNAL_FIXTURES_REGEN") == "1";

    public SignalFixtureTests(ITestOutputHelper output)
    {
        _output = output;
    }

    [Fact]
    public void X3DH_Basic_MatchesFixture()
    {
        var (inputs, expected) = LoadFixture("x3dh_basic");

        var aliceIkPriv = HexToBytes(inputs.GetProperty("alice_identity_priv_hex").GetString()!);
        var aliceEkPriv = HexToBytes(inputs.GetProperty("alice_ephemeral_priv_hex").GetString()!);
        var bobIkPriv = HexToBytes(inputs.GetProperty("bob_identity_priv_hex").GetString()!);
        var bobSpkPriv = HexToBytes(inputs.GetProperty("bob_signed_pre_key_priv_hex").GetString()!);
        var bobOpkPriv = HexToBytes(inputs.GetProperty("bob_one_time_pre_key_priv_hex").GetString()!);

        var aliceIkPub = X25519DerivePublic(aliceIkPriv);
        var aliceEkPub = X25519DerivePublic(aliceEkPriv);
        var bobIkPub = X25519DerivePublic(bobIkPriv);
        var bobSpkPub = X25519DerivePublic(bobSpkPriv);
        var bobOpkPub = X25519DerivePublic(bobOpkPriv);

        var dh1 = X25519AgreeViaService(aliceIkPriv, bobSpkPub);
        var dh2 = X25519AgreeViaService(aliceEkPriv, bobIkPub);
        var dh3 = X25519AgreeViaService(aliceEkPriv, bobSpkPub);
        var dh4 = X25519AgreeViaService(aliceEkPriv, bobOpkPub);

        var sharedSecret = Concat(dh1, dh2, dh3, dh4);
        var rootInfo = Encoding.UTF8.GetBytes(inputs.GetProperty("hkdf_root_info_utf8").GetString()!);
        var sendInfo = Encoding.UTF8.GetBytes(inputs.GetProperty("hkdf_chain_initiator_send_info_utf8").GetString()!);
        var recvInfo = Encoding.UTF8.GetBytes(inputs.GetProperty("hkdf_chain_initiator_recv_info_utf8").GetString()!);

        var rootKey = HKDF.DeriveKey(HashAlgorithmName.SHA256, sharedSecret, 32, info: rootInfo);
        var sendChainKey = HKDF.DeriveKey(HashAlgorithmName.SHA256, rootKey, 32, info: sendInfo);
        var recvChainKey = HKDF.DeriveKey(HashAlgorithmName.SHA256, rootKey, 32, info: recvInfo);

        var actuals = new Dictionary<string, byte[]>
        {
            ["alice_identity_pub_hex"] = aliceIkPub,
            ["alice_ephemeral_pub_hex"] = aliceEkPub,
            ["bob_identity_pub_hex"] = bobIkPub,
            ["bob_signed_pre_key_pub_hex"] = bobSpkPub,
            ["bob_one_time_pre_key_pub_hex"] = bobOpkPub,
            ["dh1_hex"] = dh1,
            ["dh2_hex"] = dh2,
            ["dh3_hex"] = dh3,
            ["dh4_hex"] = dh4,
            ["shared_secret_hex"] = sharedSecret,
            ["root_key_hex"] = rootKey,
            ["initiator_send_chain_key_hex"] = sendChainKey,
            ["initiator_recv_chain_key_hex"] = recvChainKey,
        };

        VerifyOrPrint("x3dh_basic", expected, actuals);
    }

    [Fact]
    public void RatchetStep_Basic_MatchesFixture()
    {
        var (inputs, expected) = LoadFixture("ratchet_step_basic");
        var chainKey = HexToBytes(inputs.GetProperty("chain_key_hex").GetString()!);

        var msgKey = HmacSha256(chainKey, 0x01);
        var nextChainKey = HmacSha256(chainKey, 0x02);

        VerifyOrPrint("ratchet_step_basic", expected, new()
        {
            ["message_key_hex"] = msgKey,
            ["next_chain_key_hex"] = nextChainKey,
        });
    }

    [Fact]
    public void RatchetStep_ThreeIterations_MatchesFixture()
    {
        var (inputs, expected) = LoadFixture("ratchet_step_three_iterations");
        var chainKey = HexToBytes(inputs.GetProperty("initial_chain_key_hex").GetString()!);

        var actuals = new Dictionary<string, byte[]>();
        for (var i = 0; i < 3; i++)
        {
            var msgKey = HmacSha256(chainKey, 0x01);
            var next = HmacSha256(chainKey, 0x02);
            actuals[$"step_{i}_message_key_hex"] = msgKey;
            actuals[$"step_{i}_chain_key_after_hex"] = next;
            chainKey = next;
        }

        VerifyOrPrint("ratchet_step_three_iterations", expected, actuals);
    }

    private void VerifyOrPrint(string caseName, JsonElement expected, Dictionary<string, byte[]> actuals)
    {
        if (RegenerateMode)
        {
            _output.WriteLine($"=== {caseName} (regen mode) ===");
            foreach (var kv in actuals)
                _output.WriteLine($"  \"{kv.Key}\": \"{Hex(kv.Value)}\",");
            return;
        }

        foreach (var kv in actuals)
        {
            if (!expected.TryGetProperty(kv.Key, out var elem))
                Assert.Fail($"[{caseName}] expected fixture missing field '{kv.Key}'.");
            var expectedHex = elem.GetString();
            var actualHex = Hex(kv.Value);
            Assert.True(
                string.Equals(expectedHex, actualHex, StringComparison.Ordinal),
                $"[{caseName}] {kv.Key} mismatch:\n  expected: {expectedHex}\n  actual:   {actualHex}");
        }
    }

    private static (JsonElement Inputs, JsonElement Expected) LoadFixture(string caseName)
    {
        var inputsPath = ResolveFixturePath("inputs.json");
        var expectedPath = ResolveFixturePath(Path.Combine("expected", $"{caseName}.json"));

        var inputsDoc = JsonDocument.Parse(File.ReadAllBytes(inputsPath));
        JsonElement caseElem = default;
        var found = false;
        foreach (var c in inputsDoc.RootElement.GetProperty("cases").EnumerateArray())
        {
            if (c.GetProperty("name").GetString() == caseName)
            {
                caseElem = c.Clone();
                found = true;
                break;
            }
        }
        if (!found)
            throw new InvalidOperationException($"Case '{caseName}' not in inputs.json.");

        var expectedDoc = JsonDocument.Parse(File.ReadAllBytes(expectedPath));
        return (caseElem, expectedDoc.RootElement.Clone());
    }

    private static byte[] HmacSha256(byte[] key, byte input) =>
        HMACSHA256.HashData(key, new byte[] { input });

    private static byte[] X25519AgreeViaService(byte[] priv, byte[] pub)
    {
        var t = typeof(SignalProtocolService).Assembly
            .GetType("AetherMesh.Security.Services.X25519Service")
            ?? throw new InvalidOperationException("X25519Service type not found.");
        var method = t.GetMethod("Agree", BindingFlags.Public | BindingFlags.Static)
            ?? throw new InvalidOperationException("X25519Service.Agree not found.");
        return (byte[])method.Invoke(null, new object[] { priv, pub })!;
    }

    private static byte[] X25519DerivePublic(byte[] priv)
    {
        var p = new X25519PrivateKeyParameters(priv, 0);
        return p.GeneratePublicKey().GetEncoded();
    }

    private static byte[] HexToBytes(string hex)
    {
        var bytes = new byte[hex.Length / 2];
        for (var i = 0; i < bytes.Length; i++)
            bytes[i] = Convert.ToByte(hex.Substring(i * 2, 2), 16);
        return bytes;
    }

    private static string Hex(byte[] data) => Convert.ToHexString(data).ToLowerInvariant();

    private static byte[] Concat(params byte[][] arrays)
    {
        var total = 0;
        foreach (var a in arrays) total += a.Length;
        var result = new byte[total];
        var offset = 0;
        foreach (var a in arrays)
        {
            Buffer.BlockCopy(a, 0, result, offset, a.Length);
            offset += a.Length;
        }
        return result;
    }

    private static string ResolveFixturePath(string relativePath)
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        while (dir != null && !File.Exists(Path.Combine(dir.FullName, "AetherProtocol.slnx")))
            dir = dir.Parent;
        if (dir == null)
            throw new InvalidOperationException("Repo root not found from test assembly.");
        return Path.Combine(dir.FullName, "fixtures", "signal", relativePath);
    }
}

File: rust/tests/signal_fixture_tests.rs
// SPDX-License-Identifier: MIT
//
// Cross-language Signal-protocol fixture verifier.
//
// Verifies that the Rust implementation produces byte-identical X3DH and
// ratchet outputs to the C# reference (committed in
// fixtures/signal/expected/*.json). Any drift between Rust and the other
// languages surfaces here as a hex mismatch.

use hkdf::Hkdf;
use hmac::{Hmac, Mac};
use serde_json::Value;
use sha2::Sha256;
use std::fs;
use std::path::PathBuf;
use x25519_dalek::{PublicKey as X25519PublicKey, StaticSecret};

type HmacSha256 = Hmac<Sha256>;

fn repo_root() -> PathBuf {
    let mut p = PathBuf::from(env!("CARGO_MANIFEST_DIR")); // .../aether-protocol/rust
    while !p.join("AetherNetProtocol.slnx").is_file() {
        if !p.pop() {
            panic!("AetherNetProtocol.slnx not found above CARGO_MANIFEST_DIR");
        }
    }
    p
}

fn load_fixture_pair(case_name: &str) -> (Value, Value) {
    let root = repo_root();
    let inputs_path = root.join("fixtures/signal/inputs.json");
    let expected_path = root.join(format!("fixtures/signal/expected/{}.json", case_name));

    let inputs: Value = serde_json::from_str(&fs::read_to_string(inputs_path).unwrap()).unwrap();
    let cases = inputs["cases"].as_array().unwrap();
    let case = cases
        .iter()
        .find(|c| c["name"].as_str() == Some(case_name))
        .unwrap()
        .clone();
    let expected: Value = serde_json::from_str(&fs::read_to_string(expected_path).unwrap()).unwrap();
    (case, expected)
}

fn unhex(s: &str) -> Vec<u8> {
    (0..s.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&s[i..i + 2], 16).unwrap())
        .collect()
}

fn hex(b: &[u8]) -> String {
    b.iter().map(|x| format!("{:02x}", x)).collect()
}

fn x25519_derive_pub(priv_bytes: &[u8]) -> [u8; 32] {
    let mut arr = [0u8; 32];
    arr.copy_from_slice(priv_bytes);
    X25519PublicKey::from(&StaticSecret::from(arr)).to_bytes()
}

fn x25519_agree(priv_bytes: &[u8], pub_bytes: &[u8]) -> [u8; 32] {
    let mut p = [0u8; 32];
    p.copy_from_slice(priv_bytes);
    let mut q = [0u8; 32];
    q.copy_from_slice(pub_bytes);
    StaticSecret::from(p)
        .diffie_hellman(&X25519PublicKey::from(q))
        .to_bytes()
}

fn hkdf32(ikm: &[u8], info: &[u8]) -> Vec<u8> {
    let hk = Hkdf::<Sha256>::new(None, ikm);
    let mut out = vec![0u8; 32];
    hk.expand(info, &mut out).unwrap();
    out
}

fn hmac_one(key: &[u8], b: u8) -> Vec<u8> {
    let mut mac = HmacSha256::new_from_slice(key).unwrap();
    mac.update(&[b]);
    mac.finalize().into_bytes().to_vec()
}

#[test]
fn signal_fixture_x3dh_basic() {
    let (inputs, expected) = load_fixture_pair("x3dh_basic");

    let alice_ik = unhex(inputs["alice_identity_priv_hex"].as_str().unwrap());
    let alice_ek = unhex(inputs["alice_ephemeral_priv_hex"].as_str().unwrap());
    let bob_ik = unhex(inputs["bob_identity_priv_hex"].as_str().unwrap());
    let bob_spk = unhex(inputs["bob_signed_pre_key_priv_hex"].as_str().unwrap());
    let bob_opk = unhex(inputs["bob_one_time_pre_key_priv_hex"].as_str().unwrap());

    let alice_ik_pub = x25519_derive_pub(&alice_ik);
    let alice_ek_pub = x25519_derive_pub(&alice_ek);
    let bob_ik_pub = x25519_derive_pub(&bob_ik);
    let bob_spk_pub = x25519_derive_pub(&bob_spk);
    let bob_opk_pub = x25519_derive_pub(&bob_opk);

    let dh1 = x25519_agree(&alice_ik, &bob_spk_pub);
    let dh2 = x25519_agree(&alice_ek, &bob_ik_pub);
    let dh3 = x25519_agree(&alice_ek, &bob_spk_pub);
    let dh4 = x25519_agree(&alice_ek, &bob_opk_pub);

    let mut shared = Vec::new();
    shared.extend_from_slice(&dh1);
    shared.extend_from_slice(&dh2);
    shared.extend_from_slice(&dh3);
    shared.extend_from_slice(&dh4);

    let root_info = inputs["hkdf_root_info_utf8"].as_str().unwrap().as_bytes();
    let send_info = inputs["hkdf_chain_initiator_send_info_utf8"]
        .as_str()
        .unwrap()
        .as_bytes();
    let recv_info = inputs["hkdf_chain_initiator_recv_info_utf8"]
        .as_str()
        .unwrap()
        .as_bytes();

    let root_key = hkdf32(&shared, root_info);
    let send_chain = hkdf32(&root_key, send_info);
    let recv_chain = hkdf32(&root_key, recv_info);

    assert_eq!(hex(&alice_ik_pub), expected["alice_identity_pub_hex"].as_str().unwrap());
    assert_eq!(hex(&alice_ek_pub), expected["alice_ephemeral_pub_hex"].as_str().unwrap());
    assert_eq!(hex(&bob_ik_pub), expected["bob_identity_pub_hex"].as_str().unwrap());
    assert_eq!(hex(&bob_spk_pub), expected["bob_signed_pre_key_pub_hex"].as_str().unwrap());
    assert_eq!(hex(&bob_opk_pub), expected["bob_one_time_pre_key_pub_hex"].as_str().unwrap());
    assert_eq!(hex(&dh1), expected["dh1_hex"].as_str().unwrap());
    assert_eq!(hex(&dh2), expected["dh2_hex"].as_str().unwrap());
    assert_eq!(hex(&dh3), expected["dh3_hex"].as_str().unwrap());
    assert_eq!(hex(&dh4), expected["dh4_hex"].as_str().unwrap());
    assert_eq!(hex(&shared), expected["shared_secret_hex"].as_str().unwrap());
    assert_eq!(hex(&root_key), expected["root_key_hex"].as_str().unwrap());
    assert_eq!(
        hex(&send_chain),
        expected["initiator_send_chain_key_hex"].as_str().unwrap()
    );
    assert_eq!(
        hex(&recv_chain),
        expected["initiator_recv_chain_key_hex"].as_str().unwrap()
    );
}

#[test]
fn signal_fixture_ratchet_step_basic() {
    let (inputs, expected) = load_fixture_pair("ratchet_step_basic");
    let chain_key = unhex(inputs["chain_key_hex"].as_str().unwrap());
    assert_eq!(
        hex(&hmac_one(&chain_key, 0x01)),
        expected["message_key_hex"].as_str().unwrap()
    );
    assert_eq!(
        hex(&hmac_one(&chain_key, 0x02)),
        expected["next_chain_key_hex"].as_str().unwrap()
    );
}

#[test]
fn signal_fixture_ratchet_step_three_iterations() {
    let (inputs, expected) = load_fixture_pair("ratchet_step_three_iterations");
    let mut chain_key = unhex(inputs["initial_chain_key_hex"].as_str().unwrap());
    for i in 0..3 {
        let msg = hmac_one(&chain_key, 0x01);
        let nxt = hmac_one(&chain_key, 0x02);
        assert_eq!(
            hex(&msg),
            expected[format!("step_{}_message_key_hex", i)].as_str().unwrap()
        );
        assert_eq!(
            hex(&nxt),
            expected[format!("step_{}_chain_key_after_hex", i)].as_str().unwrap()
        );
        chain_key = nxt;
    }
}
